README.md
# Dict_19127186

history.txt
<3
RNG
LOL
#1


src/vn/edu/hcmus/student/sv19127186/Dictionary/Game.java
package vn.edu.hcmus.student.sv19127186.Dictionary;

import javax.swing.*;
import java.awt.*;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.util.Random;
import java.util.Vector;

/**
 * vn.edu.hcmus.student.sv19127186.Dictionary
 * Created by 84904
 * Date 16/12/2021 - 11:25 PM
 * Description: ...
 */
public class Game extends JFrame implements ActionListener {
    private JLabel question;
    private JButton ans_A;
    private JButton ans_B;
    private JButton ans_C;
    private JButton ans_D;
    private int type;

    private SlangWords dict;
    Game(SlangWords temp){
        dict = temp;
        setDefaultLookAndFeelDecorated(true);
        setDefaultCloseOperation(JFrame.HIDE_ON_CLOSE);
        setTitle("Game");
        setMinimumSize(new Dimension(400,200));
        JPanel panel = new JPanel(new BorderLayout());
        //
        question = new JLabel("Question:");
        ans_A = new JButton();

        ans_B = new JButton();
        ans_C = new JButton();
        ans_D = new JButton();

        //
        ans_A.setPreferredSize(new Dimension(141, 26));
        ans_D.setPreferredSize(new Dimension(141, 26));
        ans_C.setPreferredSize(new Dimension(141, 26));
        ans_B.setPreferredSize(new Dimension(141, 26));


        //
        JPanel header = new JPanel();
        header.setLayout(new FlowLayout());
        header.add(question);

        //
        JPanel content = new JPanel();
        content.setLayout(new GridBagLayout());
        GridBagConstraints gbc = new GridBagConstraints();
        gbc.insets = new Insets(5,5,5,5);
        gbc.gridx = 0;
        gbc.gridy=0;

        content.add(ans_A,gbc);
        gbc.gridx++;
        content.add(ans_B,gbc);
        gbc.gridx=0;
        gbc.gridy++;
        content.add(ans_C,gbc);
        gbc.gridx++;
        content.add(ans_D,gbc);

        add(header,BorderLayout.PAGE_START);
        add(content,BorderLayout.CENTER);



        this.pack();
        setSize(new Dimension(400,200));
        setVisible(true);
        ans_D.addActionListener(this);
        ans_C.addActionListener(this);
        ans_B.addActionListener(this);
        ans_A.addActionListener(this);
    }

    public void run_difi(){
        type = 0;//game theo difinition
        String temp = dict.random_difinition();

        String ques = "Chose the slag word of this difinition: "+temp;
        question.setText(ques);

        Vector<String> ans = new Vector<String>();
        Vector<String> correct = dict.find_byDefinition(temp);
        String[] correct_ = correct.get(0).split("-");

        Random ran = new Random();
        int index = ran.nextInt(correct.size());

        ans.add(correct_[0]);
        for(int i =0;i<3;i++){
            ans.add(dict.random_slangword());
        }
        int tmp = ran.nextInt(4);
        temp = ans.get(0);
        String swap = ans.get(tmp);
        ans.set(0,swap);
        ans.set(tmp,temp);
        ans_A.setText(ans.get(0));
        ans_B.setText(ans.get(1));
        ans_C.setText(ans.get(2));
        ans_D.setText(ans.get(3));

        if(tmp==0){
            ans_A.setActionCommand("CORRECT");
            ans_B.setActionCommand("B");
            ans_C.setActionCommand("C");
            ans_D.setActionCommand("D");
        }
        else if(tmp ==1){
            ans_B.setActionCommand("CORRECT");
            ans_A.setActionCommand("A");
            ans_C.setActionCommand("C");
            ans_D.setActionCommand("D");
        }
        else if (tmp ==2){
            ans_C.setActionCommand("CORRECT");
            ans_A.setActionCommand("A");
            ans_B.setActionCommand("B");
            ans_D.setActionCommand("D");
        }
        else if(tmp ==3){
            ans_D.setActionCommand("CORRECT");
            ans_A.setActionCommand("A");
            ans_B.setActionCommand("B");
            ans_C.setActionCommand("C");
        }

    }

    public void run_slang(){
        type =1; //1 la game theo slang word
        String temp = dict.random_slangword();
        System.out.println(temp);
        String ques = "Chose the difinition of this word: "+temp;
        question.setText(ques);

        Vector<String> ans = new Vector<String>();
        Vector<String> correct = dict.find(temp,1);
        System.out.println(correct);
        Random ran = new Random();
        int index = ran.nextInt(correct.size());
        System.out.println(index);
        ans.add(correct.get(index));
        for(int i =0;i<3;i++){
            ans.add(dict.random_difinition());
        }
        int tmp = ran.nextInt(4);
        temp = ans.get(0);
        String swap = ans.get(tmp);
        ans.set(0,swap);
        ans.set(tmp,temp);
        ans_A.setText(ans.get(0));
        ans_B.setText(ans.get(1));
        ans_C.setText(ans.get(2));
        ans_D.setText(ans.get(3));
        System.out.println(tmp);
        if(tmp==0){
            ans_A.setActionCommand("CORRECT");
            ans_B.setActionCommand("B");
            ans_C.setActionCommand("C");
            ans_D.setActionCommand("D");
        }
        else if(tmp ==1){
            ans_B.setActionCommand("CORRECT");
            ans_A.setActionCommand("A");
            ans_C.setActionCommand("C");
            ans_D.setActionCommand("D");
        }
        else if (tmp ==2){
            ans_C.setActionCommand("CORRECT");
            ans_A.setActionCommand("A");
            ans_B.setActionCommand("B");
            ans_D.setActionCommand("D");
        }
        else if(tmp ==3){
            ans_D.setActionCommand("CORRECT");
            ans_A.setActionCommand("A");
            ans_B.setActionCommand("B");
            ans_C.setActionCommand("C");
        }

    }



    @Override
    public void actionPerformed(ActionEvent e) {
        String command =e.getActionCommand();
        if(command =="A" ){
            JOptionPane.showMessageDialog(null,"Wrong answer","Error Message", JOptionPane.ERROR_MESSAGE);
        }
        else if (command=="B"||command=="C"||command=="D"){
            JOptionPane.showMessageDialog(null,"Wrong answer","Error Message", JOptionPane.ERROR_MESSAGE);
        }
        else if (command =="CORRECT"){
            if(type==1)
                this.run_slang();
            else
                this.run_difi();
        }
    }

}


src/vn/edu/hcmus/student/sv19127186/Dictionary/add_UI.java
package vn.edu.hcmus.student.sv19127186.Dictionary;

import javax.swing.*;
import java.awt.*;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.util.Vector;

import vn.edu.hcmus.student.sv19127186.Dictionary.*;
/**
 * vn.edu.hcmus.student.sv19127186.Dictionary
 * Created by 84904
 * Date 15/12/2021 - 10:35 PM
 * Description: ...
 */
class add_nofication extends JFrame implements ActionListener{
    private SlangWords dict;
    private JButton duplicate;
    private JButton overwrite;
    private JList list;
    private String slangword_;
    private String mean_;

    add_nofication(SlangWords temp,String slangword,String mean){
        dict=temp;
        slangword_=slangword;
        mean_=mean;
        setDefaultLookAndFeelDecorated(true);
        setTitle("Notification");
        setDefaultCloseOperation(JFrame.HIDE_ON_CLOSE);
        setMinimumSize(new Dimension(300,200));
        //center
        list= new JList();
        JPanel panel = new JPanel(new BorderLayout());
        Vector<String> str = dict.find(slangword,0);
        list.setListData(str);

        //header
        JPanel header = new JPanel(new FlowLayout());
        JLabel name = new JLabel("This word has existed already", SwingConstants.CENTER);
        name.setForeground(Color.red);
        name.setFont(header.getFont().deriveFont (20.0f));
        header.add(name);

        //footer
        duplicate = new JButton("NEW");
        overwrite = new JButton("OVERWRITE");
        duplicate.setActionCommand("NEW");
        overwrite.setActionCommand("OVERWRITE");

        JPanel footerpanel = new JPanel();
        footerpanel.setLayout(new BoxLayout(footerpanel,BoxLayout.LINE_AXIS));
        footerpanel.add(duplicate);
        footerpanel.add(Box.createRigidArea(new Dimension(20,0)));
        footerpanel.add(overwrite);
        JPanel tmp = new JPanel(new BorderLayout());
        tmp.add(footerpanel,BorderLayout.LINE_END);
        JPanel panel1 = new JPanel(new FlowLayout());
        panel1.add(tmp);

        panel.add(header,BorderLayout.PAGE_START);
        panel.add(new JScrollPane(list),BorderLayout.CENTER);
        panel.add(panel1,BorderLayout.SOUTH);
        add(panel);

        this.pack();
        setSize(new Dimension(300,200));
        setVisible(true);
        duplicate.addActionListener(this);
        overwrite.addActionListener(this);
    }

    @Override
    public void actionPerformed(ActionEvent e) {
        String command =e.getActionCommand();
        if(command=="NEW"){
            dict.add_newWord(slangword_,mean_);
            JOptionPane.showMessageDialog(null,"Add successfully");
            dict.export_file("data.txt");
            this.hide();
        }
        else if(command=="OVERWRITE"){
            int index = list.getSelectedIndex();
            dict.edit_Word(slangword_,index,mean_);
            JOptionPane.showMessageDialog(null,"Add successfully");
            dict.export_file("data.txt");
            this.hide();

        }
    }
}
public class add_UI extends JFrame implements ActionListener {
    private JTextField slangword;
    private JTextField meanings;

    private JButton okbtn;
    private JButton cancelbtn;
    SlangWords dict;
    private add_nofication extra;
    add_UI(SlangWords temp){
        dict=temp;
        setDefaultLookAndFeelDecorated(true);
        setTitle("ADD");
        setDefaultCloseOperation(JFrame.HIDE_ON_CLOSE);
        setMinimumSize(new Dimension(300,200));
        JPanel panel = new JPanel(new BorderLayout());


        slangword=new JTextField(15);
        meanings = new JTextField(15);
        okbtn = new JButton("OK");
        cancelbtn = new JButton("Cancel");
        okbtn.setActionCommand("OK");
        cancelbtn.setActionCommand("Cancel");

        JPanel panel1 = new JPanel(new GridBagLayout());
        GridBagConstraints gbc = new GridBagConstraints();
        gbc.gridx=0;
        gbc.gridy=0;
        gbc.fill = GridBagConstraints.HORIZONTAL;
        gbc.anchor= GridBagConstraints.WEST;
        gbc.insets = new Insets(5,5,5,5);
        panel1.add(new JLabel("Slang Word: "),gbc);
        gbc.gridy++;
        panel1.add(new JLabel("Meanings: "),gbc);

        gbc.gridx++;
        gbc.gridy=0;
        gbc.anchor=GridBagConstraints.WEST;
        panel1.add(slangword,gbc);
        gbc.gridy++;
        panel1.add(meanings,gbc);


        //footer
        JPanel footerpanel = new JPanel();
        footerpanel.setLayout(new BoxLayout(footerpanel,BoxLayout.LINE_AXIS));
        footerpanel.add(okbtn);
        footerpanel.add(Box.createRigidArea(new Dimension(20,0)));
        footerpanel.add(cancelbtn);
        JPanel tmp = new JPanel(new BorderLayout());
        tmp.add(footerpanel,BorderLayout.LINE_END);

        JPanel panel2 = new JPanel(new FlowLayout());
        panel2.add(tmp);

        panel.add(panel1,BorderLayout.CENTER);
        panel.add(panel2,BorderLayout.SOUTH);
        add(panel);

        okbtn.addActionListener(this);
        cancelbtn.addActionListener(this);

        this.pack();
        setSize(new Dimension(300,200));
        setVisible(true);
    }

    @Override
    public void actionPerformed(ActionEvent e) {
        String command=e.getActionCommand();
        if(command=="OK"){
            String slangword_= this.slangword.getText();
            if(slangword.getText().length()==0){
                JOptionPane.showMessageDialog(null,"Please your slangword");

            }
            String mean_ = this.meanings.getText();
            if(meanings.getText().length()==0){
                JOptionPane.showMessageDialog(null,"Please input your meanings");
                return;
            }
            Vector<String>temp = dict.find(slangword_,0);
            if(temp!=null)
                extra = new add_nofication(dict,slangword_,mean_);
            else{
                dict.add_newWord(slangword_,mean_);
                JOptionPane.showMessageDialog(null,"Add successfully");
                dict.export_file("data.txt");
            }

            this.hide();

        }
        if(command=="Cancel"){
            this.hide();
            return;
        }
    }
}


src/vn/edu/hcmus/student/sv19127186/Dictionary/edit_UI.java
package vn.edu.hcmus.student.sv19127186.Dictionary;

import javax.swing.*;
import java.awt.*;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.util.Vector;

import vn.edu.hcmus.student.sv19127186.Dictionary.*;
/**
 * vn.edu.hcmus.student.sv19127186.Dictionary
 * Created by 84904
 * Date 15/12/2021 - 10:35 PM
 * Description: ...
 */
class edit_nofication extends JFrame implements ActionListener {
    private SlangWords dict;
    private JButton okbtn;
    private JButton cancel;
    private JList list;
    private String slangword_;
    private String mean_;

    edit_nofication(SlangWords temp,String slangword,String mean){
        dict=temp;
        slangword_=slangword;
        mean_=mean;
        setDefaultLookAndFeelDecorated(true);
        setTitle("Notification");
        setDefaultCloseOperation(JFrame.HIDE_ON_CLOSE);
        setMinimumSize(new Dimension(400,200));
        //center
        list= new JList();
        JPanel panel = new JPanel(new BorderLayout());
        Vector<String> str = dict.find(slangword,0);
        list.setListData(str);

        //header
        JPanel header = new JPanel(new FlowLayout());
        JLabel name = new JLabel("Which meaning do you want to update", SwingConstants.CENTER);
        name.setForeground(Color.red);
        name.setFont(header.getFont().deriveFont (20.0f));
        header.add(name);

        //footer
        okbtn = new JButton("OK");
        cancel = new JButton("CANCEL");
        okbtn.setActionCommand("OK");
        cancel.setActionCommand("CANCEL");

        JPanel footerpanel = new JPanel();
        footerpanel.setLayout(new BoxLayout(footerpanel,BoxLayout.LINE_AXIS));
        footerpanel.add(okbtn);
        footerpanel.add(Box.createRigidArea(new Dimension(20,0)));
        footerpanel.add(cancel);
        JPanel tmp = new JPanel(new BorderLayout());
        tmp.add(footerpanel,BorderLayout.LINE_END);
        JPanel panel1 = new JPanel(new FlowLayout());
        panel1.add(tmp);

        panel.add(header,BorderLayout.PAGE_START);
        panel.add(new JScrollPane(list),BorderLayout.CENTER);
        panel.add(panel1,BorderLayout.SOUTH);
        add(panel);

        this.pack();
        setSize(new Dimension(400,200));
        setVisible(true);
        okbtn.addActionListener(this);
        cancel.addActionListener(this);
    }

    @Override
    public void actionPerformed(ActionEvent e) {
        String command =e.getActionCommand();
        if(command=="OK"){
            int index = list.getSelectedIndex();
            dict.edit_Word(slangword_,index,mean_);
            JOptionPane.showMessageDialog(null,"Edit successfully");
            dict.export_file("data.txt");
            this.hide();
        }
        else{
            this.hide();
        }
    }
}
public class edit_UI extends JFrame implements ActionListener {
    private JTextField slangword;
    private JTextField meanings;

    private JButton okbtn;
    private JButton cancelbtn;
    SlangWords dict;
    private edit_nofication extra;
    edit_UI(SlangWords temp){
        dict=temp;
        setDefaultLookAndFeelDecorated(true);
        setTitle("UPDATE");
        setDefaultCloseOperation(JFrame.HIDE_ON_CLOSE);
        setMinimumSize(new Dimension(350,200));
        JPanel panel = new JPanel(new BorderLayout());


        slangword=new JTextField(15);
        meanings = new JTextField(15);
        okbtn = new JButton("OK");
        cancelbtn = new JButton("Cancel");
        okbtn.setActionCommand("OK");
        cancelbtn.setActionCommand("Cancel");

        JPanel panel1 = new JPanel(new GridBagLayout());
        GridBagConstraints gbc = new GridBagConstraints();
        gbc.gridx=0;
        gbc.gridy=0;
        gbc.fill = GridBagConstraints.HORIZONTAL;
        gbc.anchor= GridBagConstraints.WEST;
        gbc.insets = new Insets(5,5,5,5);
        panel1.add(new JLabel("Slang Word: "),gbc);
        gbc.gridy++;
        panel1.add(new JLabel("Meanings: "),gbc);

        gbc.gridx++;
        gbc.gridy=0;
        gbc.anchor=GridBagConstraints.WEST;
        panel1.add(slangword,gbc);
        gbc.gridy++;
        panel1.add(meanings,gbc);


        //footer
        JPanel footerpanel = new JPanel();
        footerpanel.setLayout(new BoxLayout(footerpanel,BoxLayout.LINE_AXIS));
        footerpanel.add(okbtn);
        footerpanel.add(Box.createRigidArea(new Dimension(20,0)));
        footerpanel.add(cancelbtn);
        JPanel tmp = new JPanel(new BorderLayout());
        tmp.add(footerpanel,BorderLayout.LINE_END);

        JPanel panel2 = new JPanel(new FlowLayout());
        panel2.add(tmp);

        panel.add(panel1,BorderLayout.CENTER);
        panel.add(panel2,BorderLayout.SOUTH);
        add(panel);

        okbtn.addActionListener(this);
        cancelbtn.addActionListener(this);

        this.pack();
        setSize(new Dimension(350,200));
        setVisible(true);
    }

    @Override
    public void actionPerformed(ActionEvent e) {
        String command=e.getActionCommand();
        if(command=="OK"){
            String slangword_= this.slangword.getText();
            if(slangword.getText().length()==0){
                JOptionPane.showMessageDialog(null,"Please your slangword");

            }
            String mean_ = this.meanings.getText();
            if(meanings.getText().length()==0){
                JOptionPane.showMessageDialog(null,"Please input your meanings");
                return;
            }
            Vector<String>temp = dict.find(slangword_,0);
            if(temp!=null)
                extra = new edit_nofication(dict,slangword_,mean_);
            else{
                dict.add_newWord(slangword_,mean_);
                JOptionPane.showMessageDialog(null,"Edit successfully");
                dict.export_file("data.txt");
            }

            this.hide();

        }
        if(command=="Cancel"){
            this.hide();
            return;
        }
    }
}


src/vn/edu/hcmus/student/sv19127186/Dictionary/delete_UI.java
package vn.edu.hcmus.student.sv19127186.Dictionary;

import javax.swing.*;
import java.awt.*;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.util.Vector;

import vn.edu.hcmus.student.sv19127186.Dictionary.*;
/**
 * vn.edu.hcmus.student.sv19127186.Dictionary
 * Created by 84904
 * Date 15/12/2021 - 10:35 PM
 * Description: ...
 */
class delete_nofication extends JFrame implements ActionListener{
    private SlangWords dict;
    private JButton delete_one;
    private JButton delete_all;
    private JList list;
    private String slangword_;


    delete_nofication(SlangWords temp,String slangword){
        dict=temp;
        slangword_=slangword;

        setDefaultLookAndFeelDecorated(true);
        setTitle("Notification");
        setDefaultCloseOperation(JFrame.HIDE_ON_CLOSE);
        setMinimumSize(new Dimension(500,200));
        //center
        list= new JList();
        JPanel panel = new JPanel(new BorderLayout());
        Vector<String> str = dict.find(slangword,0);
        list.setListData(str);

        //header
        JPanel header = new JPanel(new FlowLayout());
        JLabel name = new JLabel("Which one do you want to delete or you can delete all", SwingConstants.CENTER);
        name.setForeground(Color.red);
        name.setFont(header.getFont().deriveFont (20.0f));
        header.add(name);

        //footer
        delete_one = new JButton("DELETE ONE");
        delete_all = new JButton("DELETE ALL");
        delete_one.setActionCommand("ONE");
        delete_all.setActionCommand("ALL");

        JPanel footerpanel = new JPanel();
        footerpanel.setLayout(new BoxLayout(footerpanel,BoxLayout.LINE_AXIS));
        footerpanel.add(delete_one);
        footerpanel.add(Box.createRigidArea(new Dimension(20,0)));
        footerpanel.add(delete_all);
        JPanel tmp = new JPanel(new BorderLayout());
        tmp.add(footerpanel,BorderLayout.LINE_END);
        JPanel panel1 = new JPanel(new FlowLayout());
        panel1.add(tmp);

        panel.add(header,BorderLayout.PAGE_START);
        panel.add(new JScrollPane(list),BorderLayout.CENTER);
        panel.add(panel1,BorderLayout.SOUTH);
        add(panel);

        this.pack();
        setSize(new Dimension(500,200));
        setVisible(true);
        delete_one.addActionListener(this);
        delete_all.addActionListener(this);
    }

    @Override
    public void actionPerformed(ActionEvent e) {
        String command =e.getActionCommand();
        if(command=="ALL"){
            boolean check = dict.delete(slangword_);
            if (check == true){
                JOptionPane.showMessageDialog(null,"DELETE successfully");
                dict.export_file("data.txt");
                this.hide();
            }
            else{
                JOptionPane.showMessageDialog(null,"Cannot find this slang word","Error Message", JOptionPane.ERROR_MESSAGE);
            }

        }
        else if(command=="ONE"){
            int index = list.getSelectedIndex();
            dict.delete(slangword_,index);
            JOptionPane.showMessageDialog(null,"DELETE successfully");
            dict.export_file("data.txt");
            this.hide();

        }
    }
}
public class delete_UI extends JFrame implements ActionListener {
    private JTextField slangword;

    private JButton okbtn;
    private JButton cancelbtn;
    SlangWords dict;
    private delete_nofication extra;
    delete_UI(SlangWords temp){
        dict=temp;
        setDefaultLookAndFeelDecorated(true);
        setTitle("DELETE");
        setDefaultCloseOperation(JFrame.HIDE_ON_CLOSE);
        setMinimumSize(new Dimension(300,200));
        JPanel panel = new JPanel(new BorderLayout());


        slangword=new JTextField(15);
        okbtn = new JButton("OK");
        cancelbtn = new JButton("Cancel");
        okbtn.setActionCommand("OK");
        cancelbtn.setActionCommand("Cancel");

        JPanel panel1 = new JPanel(new GridBagLayout());
        GridBagConstraints gbc = new GridBagConstraints();
        gbc.gridx=0;
        gbc.gridy=0;
        gbc.fill = GridBagConstraints.HORIZONTAL;
        gbc.anchor= GridBagConstraints.WEST;
        gbc.insets = new Insets(5,5,5,5);
        panel1.add(new JLabel("Slang Word: "),gbc);


        gbc.gridx++;
        gbc.gridy=0;
        gbc.anchor=GridBagConstraints.WEST;
        panel1.add(slangword,gbc);


        //footer
        JPanel footerpanel = new JPanel();
        footerpanel.setLayout(new BoxLayout(footerpanel,BoxLayout.LINE_AXIS));
        footerpanel.add(okbtn);
        footerpanel.add(Box.createRigidArea(new Dimension(20,0)));
        footerpanel.add(cancelbtn);
        JPanel tmp = new JPanel(new BorderLayout());
        tmp.add(footerpanel,BorderLayout.LINE_END);

        JPanel panel2 = new JPanel(new FlowLayout());
        panel2.add(tmp);

        panel.add(panel1,BorderLayout.CENTER);
        panel.add(panel2,BorderLayout.SOUTH);
        add(panel);

        okbtn.addActionListener(this);
        cancelbtn.addActionListener(this);

        this.pack();
        setSize(new Dimension(300,200));
        setVisible(true);
    }

    @Override
    public void actionPerformed(ActionEvent e) {
        String command=e.getActionCommand();
        if(command=="OK"){
            String slangword_= this.slangword.getText();
            if(slangword.getText().length()==0){
                JOptionPane.showMessageDialog(null,"Please input your slangword");
            }
            Vector<String>temp = dict.find(slangword_,0);
            if(temp!=null) {
                extra = new delete_nofication(dict, slangword_);
                this.hide();
            }
            else{
                JOptionPane.showMessageDialog(null,"Cannot find this slang word","Error Message", JOptionPane.ERROR_MESSAGE);
            }

        }
        if(command=="Cancel"){
            this.hide();
            return;
        }
    }
}


src/vn/edu/hcmus/student/sv19127186/Dictionary/UI.java
package vn.edu.hcmus.student.sv19127186.Dictionary;

import javax.swing.*;
import javax.swing.event.ListSelectionEvent;
import javax.swing.event.ListSelectionListener;
import java.awt.*;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.io.IOException;
import java.sql.*;
import java.util.Vector;

/**
 * vn.edu.hcmus.student.sv19127186.Dictionary
 * Created by 84904
 * Date 14/12/2021 - 11:48 PM
 * Description: ...
 */


public class UI extends JPanel implements ActionListener{
    private JButton add;
    private JButton delete;
    private JButton update;
    private JButton play_slangword;
    private JButton play_difinition;
    private JButton reset_;
    private JButton search_slang;
    private JButton search_difi;
    private JButton history;
    private JTextField search;
    private JButton random_;
    SlangWords dict;

    private JList list;
    private add_UI add_ui;
    private edit_UI edit_ui;
    private delete_UI delete_ui;
    private Game game;
    UI()
    {

        dict=new SlangWords();
        setLayout(new BorderLayout());
        setMinimumSize(new Dimension(500,350));
        GridBagConstraints gbc = new GridBagConstraints();

        gbc.gridx=0;
        gbc.gridy=0;
        add = new JButton("add");
        add.setActionCommand("ADD");
        delete = new JButton("delete");
        delete.setActionCommand("DELETE");
        update = new JButton("update");
        update.setActionCommand("UPDATE");
        play_slangword = new JButton("play game with slangword");
        play_difinition = new JButton("play game with difinition");
        reset_ =new JButton("reset");
        reset_.setActionCommand("RESET");
        history = new JButton("History");
        history.setActionCommand("HISTORY");
        search_slang=new JButton("Search slang word");
        search_slang.setActionCommand("slang");
        search_difi=new JButton("Search difinition");
        search_difi.setActionCommand("difinition");
        random_=new JButton("Random slang");
        random_.setActionCommand("RANDOM");

        JPanel chucnang = new JPanel(new GridBagLayout());

        //
        list = new JList();
        Vector<String> words=new Vector<String>();
        list.setListData(words);

        //
        gbc= new GridBagConstraints();
        gbc.gridx=0;
        gbc.gridy=0;
        gbc.anchor= GridBagConstraints.EAST;
        gbc.insets = new Insets(5,5,5,5);
        chucnang.add(new JLabel("Search:"),gbc);

        search = new JTextField(12);
        gbc.gridx=1;
        gbc.fill = GridBagConstraints.HORIZONTAL;
        chucnang.add(search,gbc);

        gbc = new GridBagConstraints();
        gbc.anchor= GridBagConstraints.WEST;
        gbc.gridy=1;
        gbc.gridx=0;
        gbc.insets = new Insets(5,5,5,5);
        chucnang.add(search_slang,gbc);
        gbc.gridx++;
        search_difi.setPreferredSize(new Dimension(141, 26));
        chucnang.add(search_difi,gbc);

        gbc = new GridBagConstraints();
        gbc.anchor= GridBagConstraints.WEST;
        gbc.gridy=2;
        gbc.gridx=0;
        gbc.insets = new Insets(5,5,5,5);
        add.setPreferredSize(new Dimension(141, 26));
        chucnang.add(add,gbc);
        gbc.gridx++;

        update.setPreferredSize(new Dimension(141, 26));
        chucnang.add(update,gbc);

        gbc.gridy++;
        gbc.gridx=0;
        gbc.insets = new Insets(5,5,5,5);
        delete.setPreferredSize(new Dimension(141, 26));
        chucnang.add(delete,gbc);
        gbc.gridx++;
        reset_.setPreferredSize(new Dimension(141, 26));
        reset_.setActionCommand("RESET");
        chucnang.add(reset_,gbc);

        gbc.gridy++;
        gbc.gridx=0;
        history.setPreferredSize(new Dimension(141, 26));
        chucnang.add(history,gbc);
        gbc.gridx++;
        random_.setPreferredSize(new Dimension(141, 26));
        chucnang.add(random_,gbc);

        JPanel footerpanel = new JPanel();
        footerpanel.setLayout(new BoxLayout(footerpanel,BoxLayout.LINE_AXIS));
        footerpanel.add(play_slangword);
        play_slangword.setActionCommand("game_slang");
        footerpanel.add(Box.createRigidArea(new Dimension(20,0)));
        footerpanel.add(play_difinition);
        play_difinition.setActionCommand("game_difi");
        JPanel temp = new JPanel(new BorderLayout());
        temp.add(footerpanel,BorderLayout.LINE_END);

        JPanel panel = new JPanel(new FlowLayout());
        panel.add(temp);



        JPanel header = new JPanel(new FlowLayout());
        JLabel name = new JLabel("Dictionary", SwingConstants.CENTER);
        name.setForeground(Color.red);
        name.setFont(header.getFont().deriveFont (20.0f));
        header.add(name);

        add(header,BorderLayout.PAGE_START);
        add(chucnang,BorderLayout.LINE_START);
        add(new JScrollPane(list), BorderLayout.CENTER);
        add(panel,BorderLayout.SOUTH);

        search_slang.addActionListener(this);
        search_difi.addActionListener(this);
        play_slangword.addActionListener(this);
        play_difinition.addActionListener(this);
        add.addActionListener(this);
        delete.addActionListener(this);
        update.addActionListener(this);
        reset_.addActionListener(this);
        history.addActionListener(this);
        random_.addActionListener(this);
    }
    public void createGUI()
    {
        JFrame.setDefaultLookAndFeelDecorated(true);
        JFrame frame = new JFrame("Dictionary");
        frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);

        this.setOpaque(true);
        frame.setContentPane(this);

        frame.pack();
        frame.setSize(new Dimension(700,400));

        frame.setVisible(true);
    }
    @Override
    public void actionPerformed(ActionEvent e) {
        String command = e.getActionCommand();
        if (command == "slang") {
            String str = search.getText();
            if(str.length()==0)
            {
                Vector<String> word_ = new Vector<String>();
                this.list.setListData(word_);
            }
            else{
            Vector<String> temp = dict.find(str,0);
            if(temp!=null)
            {
                list.setListData(temp);
                dict.update_history(str);
            }
            else{
                JOptionPane.showMessageDialog(null,"Cannot find this slang word","Error Message", JOptionPane.ERROR_MESSAGE);
            }}
        }
        else if(command=="difinition"){
            String str = search.getText();
            if(str.length()==0)
                return;
            Vector<String> temp = dict.find_byDefinition(str);
            list.setListData(temp);
        }
        if(command =="RESET"){
            try {
                dict.import_file("slang.txt");
                dict.export_file("data.txt");
            } catch (IOException ex) {
                ex.printStackTrace();
            }
        }
        else if(command=="ADD"){
            add_ui= new add_UI(dict);

        }
        else if(command=="DELETE"){
            delete_ui = new delete_UI(dict);
            this.search.setText("");
            String str = this.search.getText();
            if(str.length()==0)
            {
                Vector<String> word_ = new Vector<String>();
                this.list.setListData(word_);
            }

        }
        else if(command == "UPDATE"){
            edit_ui = new edit_UI(dict);
            String str = this.search.getText();
            if(str.length()==0)
            {
                Vector<String> word_ = new Vector<String>();
                this.list.setListData(word_);
            }
        }
        else if (command=="HISTORY"){
            Vector<String>temp = dict.view_history();
            if(temp!=null)
                list.setListData(temp);
        }
        else if (command=="RANDOM"){
            String slangword = dict.random_slangword();
            Vector<String>temp= dict.find(slangword,0);
            list.setListData(temp);
        }
        else if(command=="game_slang"){
            game=new Game(dict);
            game.run_slang();
        }
        else if (command=="game_difi"){
            game = new Game(dict);
            game.run_difi();
        }

    }
}